### DISCUSSION TITLE
Providing better information in the attribute `.I_` of class `floss`

### DISCUSSION THREAD

**Discussion by NimaSarajpoor** [TDAmeritrade/stumpy#816] [General]
I have been studying the module `floss` lately and I realized that attribute `.I_` can/should provide more clear information. Based on my understanding, as we update `self._T`, we shift the time series T to the left and add a new data point to the end. We also shift `P` and `I` to the left and update them accordingly. Note that shifting `I` and `P` to the left means that we are losing information about the nearest neighbor of the very first data point in `T` and their distance to each other. In other words, updated `P` represents the updated `self._T`. Thus, updated `I` should reflect updated `self._T`. One might think that `I` contains the start index of nn of the subsequences of self._T, and what if the `nn` was something that was in the past and not in the current `self._T` (and hence we should store the index that is relative to their actual position in the time series). But, I think that is incorrect. `floss` only deals with right `nn`. Therefore, logically speaking, the values of `self._I` should be in `range(0, len(self._T))`. In other words, I should see `j==self._I[i]`, where `self._T[j:j+m]` is the RIGHT nearest neighbor of `self._T[i:i+m]`. I think this way `self._I` provides a better information compared to the existing versiton. 

I do not know how the `floss` works; however, if the goal of the attribute `.I_` is to help user has access to such information, then I think the proposed approach is better as it can provide clearer information.

What do you think @seanlaw ?

**Comment by seanlaw** [comment]
@NimaSarajpoor Here are some initial thoughts:

1. From a quick scan of the code, `self.I_` is never actually used anywhere. I'm guessing that I had added it to be consistent with other classes that return `.P_` and `.I_` but there may be limited practical use. Perhaps, we should remove it from the documentation. One might re-read the original paper to see if there was anything there.
2. Given that `self.I_` is the right matrix profile index, I don't think this is possible: `what if the nn was something that was in the past and not in the current self._T`. The nearest neighbor should always be to the right.
3. From a user standpoint, I think that it is far easier to look up the absolute index rather than a relative index as you are proposing

I wouldn't change anything unless we could motivate an actual real need as it feels like we are wasting our time guessing when nobody actually uses it or knows how to use it.

**Reply by NimaSarajpoor** [reply]
>  but there may be limited practical use

I agree. Below,

https://github.com/TDAmeritrade/stumpy/blob/a4bb1e1836e76562974d6b611d1dd71f36eb3202/stumpy/floss.py#L600-L606

the first argument is computed based on `mp[:,3]`, and I think this is the only place that might be (somehow?) replaced with `self._I`.  

But, as you said, better to not waste time if there is no real use for it.


> 2\. Given that `self.I_` is the right matrix profile index, I don't think this is possible: `what if the nn was something that was in the past and not in the current self._T`. The nearest neighbor should always be to the right.

Exaclty. I think I explained my thought badly. This is the reaon beind me suggesting that we can replace global index with local index in `self._I`

> I think that it is far easier to look up the absolute index rather than a relative index as you are proposing

Now that I think more about it... that can make sense particularly if they have the full T (up to the lastly-inserted point).
